### PR TITLE
Add built-in functions to get and set highlight groups

### PR DESCRIPTION
**Pull request by yegappan**

### CODE REVIEW

File: runtime/doc/eval.txt [hunk 1]
```diff
@@ -6716,6 +6718,81 @@ hlexists({name})					*hlexists()*
...
+		Example(s): >
+			:echo hlget()
+			:echo hlget('ModeMsg')
+<
+		Can also be used as a |method|: >
+			GetName()->hlget()
+<
+							*hlset()*
+hlset({list})
+		Set the attributes of a List of highlight groups. If a
```
Comment (romainl) [inline runtime/doc/eval.txt]: "Sets"

File: runtime/doc/eval.txt [hunk 2]
```diff
@@ -6716,6 +6718,81 @@ hlexists({name})					*hlexists()*
 
 		Can also be used as a |method|: >
 			GetName()->hlexists()
+<
+							*hlget()*
+hlget([{name}])
```
Comment (romainl) [inline runtime/doc/eval.txt]: `[{name}]` gives the impression that one must pass a potentially empty list to the function, which is not the case.

Comment (lacygoill) [reply runtime/doc/eval.txt]: > [{name}] gives the impression that one must pass a potentially empty list to the function, which is not the case.

That has always bothered me as well, but that's the convention used in the help everywhere else.
`{foo}` means that `foo` is a placeholder name (i.e. not to be taken literally, but replaced with a relevant name).
And the surrounding square brackets mean that the parameter is optional.

We can read the same `[{name}]` for `getfontname()`:
https://github.com/vim/vim/blob/2446ec9b567ce2b72bd06d121f200f40bbdc8a84/runtime/doc/eval.txt#L5787

`sign_getdefined()`:
https://github.com/vim/vim/blob/2446ec9b567ce2b72bd06d121f200f40bbdc8a84/runtime/doc/sign.txt#L423

`sign_undefine()`:
https://github.com/vim/vim/blob/2446ec9b567ce2b72bd06d121f200f40bbdc8a84/runtime/doc/sign.txt#L650

Comment (yegappan) [reply runtime/doc/eval.txt]: Yes. That is the convention used in the help files. The optional argument is surrounded by square brackets.

File: runtime/doc/eval.txt [hunk 3]
```diff
@@ -6716,6 +6718,87 @@ hlexists({name})					*hlexists()*
...
+
+		Example(s): >
+			:echo hlget()
+			:echo hlget('ModeMsg')
+<
+		Can also be used as a |method|: >
+			GetName()->hlget()
+<
+hlset({list})						*hlset()*
+		Create or modify the attributes of a List of highlight groups.
```
Comment (romainl) [inline runtime/doc/eval.txt]: "Creates or modifies" to follow the "Returns" above.

**Comment by lacygoill** [comment]
To clear a highlight group with `:highlight`, we need to choose one of 3 Ex commands, depending on how it was defined.

If it's defined with its own attributes, then we need `:highlight clear`:
```vim
highlight MyGroup ctermbg=green guibg=green
highlight clear MyGroup
highlight MyGroup
```
    MyGroup        xxx cleared

If it's linked to another highlight group, then we need to link it to `NONE`:
```vim
highlight default link MyGroup ErrorMsg
highlight link MyGroup NONE
highlight MyGroup
```
    MyGroup        xxx cleared

*Actually, `:highlight clear` might work too, but only if the group was defined without the `default` argument.*

If it's defined with its own attributes *and* is linked to another highlight group, we need to first `clear` it, *then* reset its link to `NONE` (the order of the commands matters):
```vim
highlight MyGroup ctermbg=green guibg=green
highlight! default link MyGroup ErrorMsg
highlight clear MyGroup
highlight link MyGroup NONE
highlight MyGroup
```
    MyGroup        xxx cleared

That's a lot of info.  Using `hlset()` instead of `:highlight` does not simplify the operation:
```vim
highlight MyGroup ctermbg=green guibg=green
highlight! default link MyGroup ErrorMsg
call hlset([{'name': 'MyGroup', 'cleared': v:true, 'linksto': 'NONE'}])
highlight MyGroup
```
    MyGroup        xxx ctermbg=10 guibg=green

We need 2 invocations of `hlset()` to be sure the group is cleared:
```vim
highlight MyGroup ctermbg=green guibg=green
highlight! default link MyGroup ErrorMsg
call hlset([{'name': 'MyGroup', 'cleared': v:true}])
call hlset([{'name': 'MyGroup', 'linksto': 'NONE'}])
highlight MyGroup
```
    MyGroup        xxx cleared

I think the first `hlset()` invocation fails to clear the group because it first removes the link then clears it (causing the link to be restored because of `default`).  When both the `cleared` and `linksto` keys are present, wouldn't it make more sense for `hlset()` to process `cleared` *before* `linksto`?

**Comment by vim-ml** [comment]
Hi,

On Wed, Nov 10, 2021 at 2:21 PM lacygoill ***@***.***> wrote:

> To clear a highlight group with :highlight, we need to choose one of 3 Ex
> commands, depending on how it was defined.
>
> If it's defined with its own attributes, then we need :highlight clear:
>
> highlight MyGroup ctermbg=green guibg=greenhighlight clear MyGrouphighlight MyGroup
>
> MyGroup        xxx cleared
>
> If it's linked to another highlight group, then we need to link it to NONE
> :
>
> highlight default link MyGroup ErrorMsghighlight link MyGroup NONEhighlight MyGroup
>
> MyGroup        xxx cleared
>
> *Actually, :highlight clear might work too, but only if the group was
> defined without the default argument.*
>
> If it's defined with its own attributes *and* is linked to another
> highlight group, we need to first clear it, *then* reset its link to NONE
> (the order of the commands matters):
>
> highlight MyGroup ctermbg=green guibg=greenhighlight! default link MyGroup ErrorMsghighlight clear MyGrouphighlight link MyGroup NONEhighlight MyGroup
>
> MyGroup        xxx cleared
>
> That's a lot of info. Using hlset() instead of :highlight does not
> simplify the operation:
>
> highlight MyGroup ctermbg=green guibg=greenhighlight! default link MyGroup ErrorMsgcall hlset([{'name': 'MyGroup', 'cleared': v:true, 'linksto': 'NONE'}])highlight MyGroup
>
> MyGroup        xxx ctermbg=10 guibg=green
>
> We need 2 invocations of hlset() to be sure the group is cleared:
>
> highlight MyGroup ctermbg=green guibg=greenhighlight! default link MyGroup ErrorMsgcall hlset([{'name': 'MyGroup', 'cleared': v:true}])call hlset([{'name': 'MyGroup', 'linksto': 'NONE'}])highlight MyGroup
>
> MyGroup        xxx cleared
>
> I think the first hlset() invocation fails to clear the group because it
> first removes the link then clears it (causing the link to be restored
> because of default). When both the cleared and linksto keys are present,
> wouldn't it make more sense for hlset() to process cleared *before*
> linksto?
>
>
>
I have created PR #9117 to support this.

Regards,
Yegappan

**Comment by lacygoill** [comment]
Thank you very much for the quick PR, I'm going to test it as soon as I can.

---

I noticed two other possible issues for which I would like your opinion.

Issue 1: `hlset()` can't emulate the bang of `:highlight!`.
Issue 2: `hlget()` does not give any information for the `default` argument of a linked highlight group.

With regards to issue 1:
```vim
vim9script
highlight MyGroup ctermbg=green
hlset([{name: 'MyGroup', linksto: 'ErrorMsg'}])
highlight MyGroup
```
    MyGroup        xxx ctermbg=10

Notice that the link was not installed.  I think that's because `hlset()` operates like `:highlight` by default; not like `:highlight!`.  I've read [`:help hlget()`](https://vimhelp.org/eval.txt.html#hlget%28%29) and [`:help hlset()`](https://vimhelp.org/eval.txt.html#hlset%28%29), but couldn't find any key allowing us to emulate the bang.

This means that there exists an operation that `:highlight` can perform but not `hlset()`:
```vim
vim9script
highlight MyGroup ctermbg=green
highlight! link MyGroup ErrorMsg
highlight MyGroup
```
    MyGroup        xxx ctermbg=10
                       links to ErrorMsg

Shouldn't `hlset()` provide all the features of `:highlight`?

With regards to issue 2:
```vim
vim9script
[{name: 'MyGroup', linksto: 'ErrorMsg'}]->hlset()
highlight clear MyGroup
highlight MyGroup
```
    MyGroup     xxx cleared

Notice that `MyGroup` has been cleared.  This means that when `hlset()` defines a group with a link, it does not use `default`.  How are we supposed to tell `hlset()` that we do want `default`?  Shouldn't it accept a boolean key like `default_link`?

---

If you consider all of this as valid, I can open separate issues; let me know.  Thank you again for your work on these very useful functions.

**Comment by vim-ml** [comment]
Hi,

On Thu, Nov 11, 2021 at 5:20 AM lacygoill ***@***.***> wrote:

> Thank you very much for the quick PR, I'm going to test it as soon as I
> can.
> ------------------------------
>
> I noticed two other possible issues for which I would like your opinion.
>
> Issue 1: hlset() can't emulate the bang of :highlight!.
> Issue 2: hlget() does not give any information for the default argument
> of a linked highlight group.
>

Thanks for the detailed feedback. I have updated the PR to add support for
these.
Can you try the updated PR and let me know if you see any problems?

Regards,
Yegappan


> With regards to issue 1:
>
> vim9scripthighlight MyGroup ctermbg=greenhlset([{name: 'MyGroup', linksto: 'ErrorMsg'}])highlight MyGroup
>
> MyGroup        xxx ctermbg=10
>
> Notice that the link was not installed. I think that's because hlset()
> operates like :highlight by default; not like :highlight!. I've read :help
> hlget() <https://vimhelp.org/eval.txt.html#hlget%28%29> and :help hlset()
> <https://vimhelp.org/eval.txt.html#hlset%28%29>, but couldn't find any
> key allowing us to emulate the bang.
>
> This means that there exists an operation that :highlight can perform but
> not hlset():
>
> vim9scripthighlight MyGroup ctermbg=greenhighlight! link MyGroup ErrorMsghighlight MyGroup
>
> MyGroup        xxx ctermbg=10
>                    links to ErrorMsg
>
> Shouldn't hlset() provide all the features of :highlight?
>
> With regards to issue 2:
>
> vim9script
> [{name: 'MyGroup', linksto: 'ErrorMsg'}]->hlset()highlight clear MyGrouphighlight MyGroup
>
> MyGroup     xxx cleared
>
> Notice that MyGroup has been cleared. This means that when hlset()
> defines a group with a link, it does not use default. How are we supposed
> to tell hlset() that we do want default? Shouldn't it accept a boolean
> key like default_link?
> ------------------------------
>
> If you consider all of this as valid, I can open separate issues; let me
> know. Thank you again for your work on these very useful functions.
>
>
>

**Comment by lacygoill** [comment]
Thank you very much for the quick fix.  Everything works fine with maybe one exception.  The `default` key is still absent from the output of `hlget()`.
```vim
vim9script
[{name: 'MyGroup', linksto: 'ErrorMsg', default: true}]->hlset()
echo hlget('MyGroup')->get(0, {})->has_key('default')
```
    0

This matters if we want to create an exact duplicate of an existing highlight group, and maybe slightly alter one of its attribute.  For example:
```vim
vim9script
highlight default link MyGroup ErrorMsg
var hg_save: list<dict<any>> = hlget('MyGroup')
var dup: string = 'Duplicate'
hg_save[0]->extend({'name': dup})
hg_save->hlset()
highlight clear Duplicate
highlight Duplicate
```
    Duplicate     xxx cleared

Notice that `Duplicate` is cleared.  This means that when it was created by `hlset()`, the latter did not use `default`; otherwise, the link would have been automatically restored just like here:
```vim
vim9script
highlight default link MyGroup ErrorMsg
highlight clear MyGroup
highlight MyGroup
```
    MyGroup        xxx links to ErrorMsg

I think the issue comes from `hlget()` which still does not tell us whether `MyGroup` is linked to `ErrorMsg` with `default` or not.  It would help if the output of `hlget()` included `default: true` or `default: false` for a linked group:
```vim
vim9script
[{name: 'MyGroup', linksto: 'ErrorMsg', default: true}]->hlset()
echo hlget('MyGroup')
```
    [{'id': 54, 'linksto': 'ErrorMsg', 'default': true, 'name': 'MyGroup'}]
                                       ^-------------^

**Comment by vim-ml** [comment]
Hi,

On Thu, Nov 11, 2021 at 8:30 PM lacygoill ***@***.***> wrote:

> Thank you very much for the quick fix. Everything works fine with maybe
> one exception. The default key is still absent from the output of hlget().
>

The 'default' argument to the highlight command is only used to determine
whether the new
attributes/link can override the already existing attributes/link and the
'default' setting is not
stored with the highlight group in the internal Vim structures. So the
hlget() function cannot
return this as it is not present in the internal Vim highlight group
structures.

For example, you can run the following commands:

:highlight hlg1 ctermfg=2
:highlight default hlg2 ctermfg=2
:highlight hlg1
hlg1           xxx ctermfg=2
:highlight hlg2
hlg2           xxx ctermfg=2

As you can see from the above output, both hlg1 and hlg2 have similar
output and the
"default" setting is missing for hlg2.

Regards,
Yegappan


> vim9script
> [{name: 'MyGroup', linksto: 'ErrorMsg', default: true}]->hlset()echo hlget('MyGroup')->get(0, {})->has_key('default')
>
> 0
>
> This matters if we want to create an exact duplicate of an existing
> highlight group, and maybe slightly alter one of its attribute. For example:
>
> vim9scripthighlight default link MyGroup ErrorMsgvar hg_save: list<dict<any>> = hlget('MyGroup')var dup: string = 'Duplicate'
> hg_save[0]->extend({'name': dup})
> hg_save->hlset()highlight clear Duplicatehighlight Duplicate
>
> Duplicate     xxx cleared
>
> Notice that Duplicate is cleared. This means that when it was created by
> hlset(), the latter did not use default; otherwise, the link would have
> been automatically restored just like here:
>
> vim9scripthighlight default link MyGroup ErrorMsghighlight clear MyGrouphighlight MyGroup
>
> MyGroup        xxx links to ErrorMsg
>
> I think the issue comes from hlget() which still does not tell us whether
> MyGroup is linked to ErrorMsg with default or not. It would help if the
> output of hlget() included default: true or default: false for a linked
> group:
>
> vim9script
> [{name: 'MyGroup', linksto: 'ErrorMsg', default: true}]->hlset()echo hlget('MyGroup')
>
> [{'id': 54, 'linksto': 'ErrorMsg', 'default': true, 'name': 'MyGroup'}]
>                                    ^-------------^
>
>
>

**Comment by vim-ml** [comment]
Hi,

On Fri, Nov 12, 2021 at 8:14 AM Yegappan Lakshmanan ***@***.***> wrote:
>
> Hi,
>
> On Thu, Nov 11, 2021 at 8:30 PM lacygoill ***@***.***> wrote:
>>
>> Thank you very much for the quick fix. Everything works fine with maybe one exception. The default key is still absent from the output of hlget().
>
>
> The 'default' argument to the highlight command is only used to determine whether the new
> attributes/link can override the already existing attributes/link and the 'default' setting is not
> stored with the highlight group in the internal Vim structures. So the hlget() function cannot
> return this as it is not present in the internal Vim highlight group structures.
>

After sending this reply, I noticed that Vim does store the 'default'
setting for links,
but doesn't store it for highlight attributes. I have updated the PR
to return the
default setting for links.

Regards,
Yegappan

> For example, you can run the following commands:
>
> :highlight hlg1 ctermfg=2
> :highlight default hlg2 ctermfg=2
> :highlight hlg1
> hlg1           xxx ctermfg=2
> :highlight hlg2
> hlg2           xxx ctermfg=2
>
> As you can see from the above output, both hlg1 and hlg2 have similar output and the
> "default" setting is missing for hlg2.
>
> Regards,
> Yegappan
>
>>
>> vim9script
>> [{name: 'MyGroup', linksto: 'ErrorMsg', default: true}]->hlset()
>> echo hlget('MyGroup')->get(0, {})->has_key('default')
>>
>> 0
>>
>> This matters if we want to create an exact duplicate of an existing highlight group, and maybe slightly alter one of its attribute. For example:
>>
>> vim9script
>> highlight default link MyGroup ErrorMsg
>> var hg_save: list<dict<any>> = hlget('MyGroup')
>> var dup: string = 'Duplicate'
>> hg_save[0]->extend({'name': dup})
>> hg_save->hlset()
>> highlight clear Duplicate
>> highlight Duplicate
>>
>> Duplicate     xxx cleared
>>
>> Notice that Duplicate is cleared. This means that when it was created by hlset(), the latter did not use default; otherwise, the link would have been automatically restored just like here:
>>
>> vim9script
>> highlight default link MyGroup ErrorMsg
>> highlight clear MyGroup
>> highlight MyGroup
>>
>> MyGroup        xxx links to ErrorMsg
>>
>> I think the issue comes from hlget() which still does not tell us whether MyGroup is linked to ErrorMsg with default or not. It would help if the output of hlget() included default: true or default: false for a linked group:
>>
>> vim9script
>> [{name: 'MyGroup', linksto: 'ErrorMsg', default: true}]->hlset()
>> echo hlget('MyGroup')
>>
>> [{'id': 54, 'linksto': 'ErrorMsg', 'default': true, 'name': 'MyGroup'}]
>>                                    ^-------------^
>>
>>

**Comment by lacygoill** [comment]
> I have updated the PR to return the default setting for links.

Thank you very much.  I've just tested the PR, and everything works as expected.

**Comment by lacygoill** [comment]
Interestingly enough, `:highlight` is faster than `hlset()`:
```vim
vim9script
var time = reltime()
def Func()
    for i: number in range(100'000)
        highlight ErrorMsg term=standout ctermfg=253 ctermbg=168 guifg=#dadada guibg=#d75f87
    endfor
enddef
Func()
(reltime(time)->reltimestr()->matchstr('.*\..\{,3}') .. ' seconds to run Func()')->setline(1)
```
    0.343 seconds to run Func()
```vim
vim9script
var time = reltime()
def Func()
    for i: number in range(100'000)
        [{
            name: 'ErrorMsg',
            ctermfg: '253',
            ctermbg: '168',
            guibg: '#d75f87',
            guifg: '#dadada',
            term: {standout: true},
        }]->hlset()
    endfor
enddef
Func()
(reltime(time)->reltimestr()->matchstr('.*\..\{,3}') .. ' seconds to run Func()')->setline(1)
```
    0.511 seconds to run Func()

The discrepancy is more pronounced at the script level:
```vim
vim9script
var time = reltime()
for i in range(100'000)
    highlight ErrorMsg term=standout ctermfg=253 ctermbg=168 guifg=#dadada guibg=#d75f87
endfor
(reltime(time)->reltimestr()->matchstr('.*\..\{,3}') .. ' seconds to run the script')->setline(1)
```
    0.523 seconds to run the script
```vim
vim9script
var time = reltime()
for i: number in range(100'000)
    [{
        name: 'ErrorMsg',
        ctermfg: '253',
        ctermbg: '168',
        guibg: '#d75f87',
        guifg: '#dadada',
        term: {standout: true},
    }]->hlset()
endfor
(reltime(time)->reltimestr()->matchstr('.*\..\{,3}') .. ' seconds to run the script')->setline(1)
```
    1.268 seconds to run the script

[Log for `hlset()`](https://0x0.st/-RG5.txt).
[Log for `:highlight`](https://0x0.st/-RGn.txt).

**Comment by vim-ml** [comment]
Hi,

On Fri, Nov 19, 2021 at 5:27 AM lacygoill ***@***.***> wrote:

> Interestingly enough, :highlight is faster than hlset():
>
> vim9scriptvar time = reltime()def Func()
>     for i: number in range(100'000)
>         highlight ErrorMsg term=standout ctermfg=253 ctermbg=168 guifg=#dadada guibg=#d75f87
>     endforenddefFunc()
> (reltime(time)->reltimestr()->matchstr('.*\..\{,3}') .. ' seconds to run Func()')->setline(1)
>
> 0.343 seconds to run Func()
>
> vim9scriptvar time = reltime()def Func()
>     for i: number in range(100'000)
>         [{
>             name: 'ErrorMsg',
>             ctermfg: '253',
>             ctermbg: '168',
>             guibg: '#d75f87',
>             guifg: '#dadada',
>             term: {standout: true},
>         }]->hlset()
>     endforenddefFunc()
> (reltime(time)->reltimestr()->matchstr('.*\..\{,3}') .. ' seconds to run Func()')->setline(1)
>
> 0.511 seconds to run Func()
>
> The discrepancy is more pronounced at the script level:
>
> vim9scriptvar time = reltime()for i in range(100'000)
>     highlight ErrorMsg term=standout ctermfg=253 ctermbg=168 guifg=#dadada guibg=#d75f87endfor
> (reltime(time)->reltimestr()->matchstr('.*\..\{,3}') .. ' seconds to run the script')->setline(1)
>
> 0.523 seconds to run the script
>
> vim9scriptvar time = reltime()for i: number in range(100'000)
>     [{
>         name: 'ErrorMsg',
>         ctermfg: '253',
>         ctermbg: '168',
>         guibg: '#d75f87',
>         guifg: '#dadada',
>         term: {standout: true},
>     }]->hlset()endfor
> (reltime(time)->reltimestr()->matchstr('.*\..\{,3}') .. ' seconds to run the script')->setline(1)
>
> 1.268 seconds to run the script
>
>
>
I took a look at this issue. It looks like just populating the attributes
itself takes time. For example,
if you source the following script:

vim9script
var time = reltime()
var attr: list<dict<any>>
for i: number in range(100'000)
    attr = [{
        name: 'ErrorMsg',
        ctermfg: '253',
        ctermbg: '168',
        guibg: '#d75f87',
        guifg: '#dadada',
        term: {standout: true},
    }]
endfor
(reltime(time)->reltimestr()->matchstr('.*\..\{,3}') .. ' seconds to run
the script')->setline(1)

The above script takes around 0.630 seconds in my setup (without any calls
to hlset()).
So it looks like most of the time is spent outside of the hlset() function.

Regards,
Yegappan

**Comment by vim-ml** [comment]
Hi,

On Sat, Nov 20, 2021 at 11:32 AM Yegappan Lakshmanan
***@***.***> wrote:
>
> On Fri, Nov 19, 2021 at 5:27 AM lacygoill ***@***.***> wrote:
>>
>> Interestingly enough, :highlight is faster than hlset():
>>
>> vim9script
>> var time = reltime()
>> def Func()
>>     for i: number in range(100'000)
>>         highlight ErrorMsg term=standout ctermfg=253 ctermbg=168 guifg=#dadada guibg=#d75f87
>>     endfor
>> enddef
>> Func()
>> (reltime(time)->reltimestr()->matchstr('.*\..\{,3}') .. ' seconds to run Func()')->setline(1)
>>
>> 0.343 seconds to run Func()
>>
>> vim9script
>> var time = reltime()
>> def Func()
>>     for i: number in range(100'000)
>>         [{
>>             name: 'ErrorMsg',
>>             ctermfg: '253',
>>             ctermbg: '168',
>>             guibg: '#d75f87',
>>             guifg: '#dadada',
>>             term: {standout: true},
>>         }]->hlset()
>>     endfor
>> enddef
>> Func()
>> (reltime(time)->reltimestr()->matchstr('.*\..\{,3}') .. ' seconds to run Func()')->setline(1)
>>
>> 0.511 seconds to run Func()
>>
>> The discrepancy is more pronounced at the script level:
>>
>> vim9script
>> var time = reltime()
>> for i in range(100'000)
>>     highlight ErrorMsg term=standout ctermfg=253 ctermbg=168 guifg=#dadada guibg=#d75f87
>> endfor
>> (reltime(time)->reltimestr()->matchstr('.*\..\{,3}') .. ' seconds to run the script')->setline(1)
>>
>> 0.523 seconds to run the script
>>
>> vim9script
>> var time = reltime()
>> for i: number in range(100'000)
>>     [{
>>         name: 'ErrorMsg',
>>         ctermfg: '253',
>>         ctermbg: '168',
>>         guibg: '#d75f87',
>>         guifg: '#dadada',
>>         term: {standout: true},
>>     }]->hlset()
>> endfor
>> (reltime(time)->reltimestr()->matchstr('.*\..\{,3}') .. ' seconds to run the script')->setline(1)
>>
>> 1.268 seconds to run the script
>>
>>
>
> I took a look at this issue. It looks like just populating the attributes itself takes time. For example,
> if you source the following script:
>
> vim9script
> var time = reltime()
> var attr: list<dict<any>>
> for i: number in range(100'000)
>     attr = [{
>         name: 'ErrorMsg',
>         ctermfg: '253',
>         ctermbg: '168',
>         guibg: '#d75f87',
>         guifg: '#dadada',
>         term: {standout: true},
>     }]
> endfor
> (reltime(time)->reltimestr()->matchstr('.*\..\{,3}') .. ' seconds to run the script')->setline(1)
>
> The above script takes around 0.630 seconds in my setup (without any calls to hlset()).
> So it looks like most of the time is spent outside of the hlset() function.
>

If I modify the script so that the attributes are initialized outside
of the for loop, then the
script runs faster (around 0.520 seconds):

vim9script
var attr: list<dict<any>> = [{
    name: 'ErrorMsg',
    ctermfg: '253',
    ctermbg: '168',
    guibg: '#d75f87',
    guifg: '#dadada',
    term: {standout: true},
}]
var time = reltime()
for i: number in range(100'000)
    attr->hlset()
endfor
(reltime(time)->reltimestr()->matchstr('.*\..\{,3}') .. ' seconds to
run the script')->setline(1)

Similarly, if hlset() is called for a large number of attributes in a
single invocation, then
the script runs faster (0.230 seconds):

vim9script
var l: list<dict<any>> = []
for i: number in range(100'000)
    l->extend([{
        name: 'ErrorMsg',
        ctermfg: '253',
        ctermbg: '168',
        guibg: '#d75f87',
        guifg: '#dadada',
        term: {standout: true},
    }])
endfor
var time = reltime()
l->hlset()
(reltime(time)->reltimestr()->matchstr('.*\..\{,3}') .. ' seconds to
run the script')->setline(1)

Regards,
Yegappan

**Pull request closed without merging**